Java/CoreCrypto/src/main/java/com/breadwallet/corecrypto/Network.java

```diff
@@ -51,252 +63,180 @@ static Network from(com.breadwallet.crypto.Network network) {
+            if (code.equals(currency.getCode())) {
+                return Optional.of(currency);
+            }
+        }
+        return Optional.absent();
     }
 
-    /* package */
-    BRChainParams asBtc() {
-        return impl.asBtc();
+    @Override
+    public boolean hasCurrency(com.breadwallet.crypto.Currency currency) {
+        Currency currencyImpl = Currency.from(currency);
+        return core.hasCurrency(currencyImpl.getCoreBRCryptoCurrency());
     }
 
-    /* package */
-    void setHeight(UnsignedLong height) {
-        this.height = height;
+    @Override
+    public Optional<Unit> baseUnitFor(com.breadwallet.crypto.Currency currency) {
+        Currency currencyImpl = Currency.from(currency);
+        return core.getUnitAsBase(currencyImpl.getCoreBRCryptoCurrency()).transform(Unit::create);
     }
 
-    // TODO(discuss): Should asBtc() and asEth() be returning Optional instead of throwing an exception?
-    private interface CurrencyNetwork {
-        default BRChainParams asBtc() {
-            throw new IllegalStateException("Invalid network type");
-        }
-
-        List<WalletManagerMode> getSupportedModes();
-
-        Optional<com.breadwallet.crypto.Address> addressFor(String address);
+    @Override
+    public Optional<Unit> defaultUnitFor(com.breadwallet.crypto.Currency currency) {
+        Currency currencyImpl = Currency.from(currency);
+        return core.getUnitAsDefault(currencyImpl.getCoreBRCryptoCurrency()).transform(Unit::create);
     }
 
-    private static class BitcoinNetwork implements CurrencyNetwork {
-
-        private final BRChainParams chainParams;
+    @Override
+    public Optional<Set<? extends com.breadwallet.crypto.Unit>> unitsFor(com.breadwallet.crypto.Currency currency) {
+        Currency currencyImpl = Currency.from(currency);
+        CoreBRCryptoCurrency currencyCore = currencyImpl.getCoreBRCryptoCurrency();
 
-        BitcoinNetwork(boolean isMainnet) {
-            Pointer ptr = CryptoLibrary.LIBRARY.getGlobalVariableAddress(isMainnet ? "BRMainNetParams" : "BRTestNetParams");
-            checkNotNull(ptr);
-            this.chainParams = new BRChainParams(ptr.getPointer(0));
-        }
+        Set<Unit> units = new HashSet<>();
+        UnsignedLong count = core.getUnitCount(currencyCore);
 
-        @Override
-        public BRChainParams asBtc() {
-            return chainParams;
-        }
+        for (UnsignedLong i = UnsignedLong.ZERO; i.compareTo(count) < 0; i = i.plus(UnsignedLong.ONE)) {
+            Optional<Unit> unit = core.getUnitAt(currencyCore, i).transform(Unit::create);
+            if (!unit.isPresent()) {
+                return Optional.absent();
+            }
 
-        @Override
-        public List<WalletManagerMode> getSupportedModes() {
-            return Collections.singletonList(WalletManagerMode.P2P_ONLY);
+            units.add(unit.get());
         }
+        return units.isEmpty() ? Optional.absent() : Optional.of(units);
+    }
```

Comment:
The current implementation is actually consistent with the Swift. Originally, I had it how you've proposed but the unit tests (they did their job!) highlighted that the Swift behaves this way.

Question now is; what do we actually want?

Reply:
```
    public func unitsFor (currency: Currency) -> Set<Unit>? {
        guard hasCurrency (currency) else { return nil }
        return Set ((0..<cryptoNetworkGetUnitCount (core, currency.core))
            .map { cryptoNetworkGetUnitAt (core, currency.core, $0) }
            .map { Unit (core: $0, take: false) }
        )
    }
```
I think it has switched back and forth.  The above is what the Swift currently is: absent if currency is unknown, otherwise a possibly empty Set.

Reply:
Ah, I see, there's the subtlety of a currency being there but not units associated with it?

Reply:
Yes, at least possibly.